sdk/tables/azure-data-tables/tests/test_table_client_cosmos.py

```diff
@@ -175,6 +183,318 @@ def test_table_name_errors_bad_length(self, tables_cosmos_account_name, tables_p
+        assert ex_msg == str(ex.value)
+
+    @cosmos_decorator
+    @recorded_by_proxy
+    def test_table_service_client_with_sas_token_credential(
+        self, tables_cosmos_account_name, tables_primary_cosmos_account_key
+    ):
+        base_url = self.account_url(tables_cosmos_account_name, "cosmos")
+        table_name = self.get_resource_name("mytable")
+        sas_token = self.generate_sas(
+            generate_account_sas,
+            tables_primary_cosmos_account_key,
+            resource_types=ResourceTypes.from_string("sco"),
+            permission=AccountSasPermissions.from_string("rwdlacu"),
+            expiry=datetime.utcnow() + timedelta(hours=1),
+        )
+        name_filter = "TableName eq '{}'".format(table_name)
+
+        with TableServiceClient(base_url, credential=AzureSasCredential(sas_token)) as client:
+            client.create_table(table_name)
+            result = client.query_tables(name_filter)
+            assert len(list(result)) == 1
+
+        conn_str = (
+            f"AccountName={tables_cosmos_account_name};SharedAccessSignature={AzureSasCredential(sas_token).signature}"
+        )
+        with TableServiceClient.from_connection_string(conn_str) as client:
+            result = client.query_tables(name_filter)
+            with pytest.raises(ClientAuthenticationError) as ex:
+                for table in result:
+                    pass
+            ex_msg = "Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature."
+            assert ex_msg in str(ex.value)
+
+        with TableServiceClient(base_url, credential=AzureSasCredential(sas_token)) as client:
+            client.delete_table(table_name)
+
+        with pytest.raises(ValueError) as ex:
+            TableServiceClient(f"{base_url}/?{sas_token}", credential=AzureSasCredential(sas_token))
+        ex_msg = "You cannot use AzureSasCredential when the resource URI also contains a Shared Access Signature."
+        assert ex_msg == str(ex.value)
+
+    @cosmos_decorator
+    @recorded_by_proxy
+    def test_table_client_with_token_credential(self, tables_cosmos_account_name, tables_primary_cosmos_account_key):
+        # DefaultAzureCredential doesn't work on Cosmos
+        base_url = self.account_url(tables_cosmos_account_name, "cosmos")
+        table_name = self.get_resource_name("mytable")
+        default_azure_credential = self.get_token_credential()
+        sas_token = self.generate_sas(
+            generate_account_sas,
+            tables_primary_cosmos_account_key,
+            resource_types=ResourceTypes.from_string("sco"),
+            permission=AccountSasPermissions.from_string("rwdlacu"),
+            expiry=datetime.utcnow() + timedelta(hours=1),
+        )
+
+        with TableClient(base_url, table_name, credential=default_azure_credential) as client:
+            with pytest.raises(HttpResponseError) as ex:
```

Comment:
hmm this should probably be a ClientAuthenticationError. But that fine, we can fix that as a non-breaking change during the refactor.

Reply:
After investigating, yes, it should be `ClientAuthenticationError`. It's a bug in `_decode_error()`. The error type is `ClientAuthenticationError` before decode with error_code "Unauthorized", but the error_code is categorized as `HttpResponseError` in these condition checks:
https://github.com/Azure/azure-sdk-for-python/blob/9156458ce01a6325af78a746d52e49e1d7d3a38d/sdk/tables/azure-data-tables/azure/data/tables/_error.py#L147-L171

Reply:
Cool - we can fix this later in the refactor.